text/0000-cargo-cfg-dependencies.md

```diff
@@ -0,0 +1,168 @@
+- Feature Name: N/A
+- Start Date: 2015-11-10
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Improve the target-specific dependency experience in Cargo by leveraging the
+same `#[cfg]` syntax that Rust has.
+
+# Motivation
+[motivation]: #motivation
+
+Currently in Cargo it's [relatively painful][issue] to list target-specific
+dependencies. This can only be done by listing out the entire target string as
+opposed to using the more-convenient `#[cfg]` annotations that Rust source code
+has access to. Consequently a Windows-specific dependency ends up having to be
+defined for four triples: `{i686,x86_64}-pc-windows-{gnu,msvc}`, and this is
+unfortunately not forwards compatible as well!
+
+[issue]: https://github.com/rust-lang/cargo/issues/1007
+
+As a result most crates end up unconditionally depending on target-specific
+dependencies and rely on the crates themselves to have the relevant `#[cfg]` to
+only be compiled for the right platforms. This experience leads to excessive
+downloads, excessive compilations, and overall "unclean methods" to have a
+platform specific dependency.
+
+This RFC proposes leveraging the same familiar syntax used in Rust itself to
+define these dependencies.
+
+# Detailed design
+[design]: #detailed-design
+
+The target-specific dependency syntax in Cargo will be expanded to to include
+not only full target strings but also `#[cfg]` expressions:
+
+```toml
+[target."cfg(windows)".dependencies]
+winapi = "0.2"
+
+[target."cfg(unix)".dependencies]
+unix-socket = "0.4"
+
+[target."cfg(target_os = \"macos\")".dependencies]
+core-foundation = "0.2"
+```
+
+Specifically, the "target" listed here is considered special if it starts with
+the string "cfg(" and ends with ")". If this is not true then Cargo will
+continue to treat it as an opaque string and pass it to the compiler via
+`--target` (Cargo's current behavior).
+
+> **Note**: There's an [issue open against TOML][toml-issue] to support
+> single-quoted keys allowing more ergonomic syntax in some cases like:
```

Comment:
In reading the spec, however, there's a difference between a string literal and a "basic string", e.g. the term "basic string" is used to reference strings that are surrounded by double quotes, and the term for strings surrounded by single quotes is "literal string". This interpretation is also confirmed by the [proposed ABNF](https://github.com/toml-lang/toml/pull/236) where the `key` production only allows for quotation-mark-surrounded strings.

Now this may all just be a misunderstanding, though, as it may be intended that literal (single-quote) strings are allowed in keys as well